api/client/charms/localcharmputters.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charms
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+
+	"github.com/juju/charm/v12"
+	"github.com/juju/errors"
+	"gopkg.in/httprequest.v1"
+
+	"github.com/juju/juju/api/base"
+	"github.com/juju/juju/rpc/params"
+)
+
+// CharmPutter uploads a local charm blob to the controller
+type CharmPutter interface {
+	PutCharm(ctx context.Context, modelUUID, charmRef, curl string, body io.Reader) (string, error)
+}
+
+// httpPutter uploads local charm blobs to the controller via the legacy
+// "model/:modeluuid/charms" endpoint hosted by the controller
+type httpPutter struct {
+	httpClient *httprequest.Client
+}
+
+// newHTTPPutter create an httpPutter, which uploads local charm blobs
+// to the controller via the legacy "model/:modeluuid/charms" endpoint
+// hosted by the controller
+func newHTTPPutter(api base.APICaller) (CharmPutter, error) {
+	// The returned httpClient sets the base url to /model/:modeluuid if it can.
+	apiHTTPClient, err := api.HTTPClient()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot retrieve http client from the api connection")
+	}
+	return &httpPutter{
+		httpClient: apiHTTPClient,
+	}, nil
+}
+
+func (h *httpPutter) PutCharm(ctx context.Context, _, _, curlStr string, body io.Reader) (string, error) {
+	curl, err := charm.ParseURL(curlStr)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	args := url.Values{}
+	args.Add("series", curl.Series)
+	args.Add("schema", curl.Schema)
+	args.Add("revision", strconv.Itoa(curl.Revision))
+	apiURI := url.URL{Path: "/charms", RawQuery: args.Encode()}
+
+	var resp params.CharmsResponse
+	req, err := http.NewRequest("POST", apiURI.String(), body)
+	if err != nil {
+		return "", errors.Annotate(err, "cannot create upload request")
+	}
+	req.Header.Set("Content-Type", "application/zip")
+
+	if err := h.httpClient.Do(ctx, req, &resp); err != nil {
+		return "", errors.Trace(err)
+	}
+	return resp.CharmURL, nil
+}
+
+// s3Putter uploads local charm blobs to the controller via the
+// s3-compatible api endpoint "model-:modeluuid/charms/:object"
+type s3Putter struct {
+	httpClient *httprequest.Client
+}
```

Comment:
Please add comments as to why the s3Putter is using an http client. It's not readily apparent as the httpPutter also uses the same http client.

Reply:
Done